pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -0,0 +1,360 @@
+			controllerutil.AddFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "subnetset", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on subnetset CR", "subnetset", req.NamespacedName)
+		}
+
+		if err := r.updateSubnetSetStatus(obj); err != nil {
+			log.Error(err, "update subnetset status failed, would retry exponentially", "subnetset", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResTypeSubnetSet)
+			if err := r.deleteSubnets(obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnetset", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			controllerutil.RemoveFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnetset", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "subnetset", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			log.Info("finalizers cannot be recognized", "subnetset", req.NamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// createSubnet creates subnet managed by subnetset
+func (r *SubnetSetReconciler) createSubnet(subnetset *v1alpha1.SubnetSet, name string) (*v1alpha1.Subnet, error) {
+	var subnet *v1alpha1.Subnet
+	if err := retry.OnError(retry.DefaultRetry, func(err error) bool {
+		return err != nil
+	}, func() error {
+		blockDeletion := true
+		obj := &v1alpha1.Subnet{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: subnetset.Namespace,
+				Name:      name,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         "v1alpha1",
+						Kind:               "subnetsets.nsx.vmware.com",
+						Name:               subnetset.Name,
+						UID:                subnetset.UID,
+						Controller:         nil,
+						BlockOwnerDeletion: &blockDeletion,
+					},
+				},
+			},
+			Spec: v1alpha1.SubnetSpec{
+				IPv4SubnetSize: subnetset.Spec.IPv4SubnetSize,
+				AccessMode:     subnetset.Spec.AccessMode,
+			},
+		}
+		if err := r.Client.Create(context.Background(), obj); err != nil {
+			return err
+		}
+		subnet = obj
+		return nil
+	}); err != nil {
+		log.Error(err, "failed to create subnet", "Namespace", subnetset.Namespace, "Name", name)
+		return nil, err
+	}
+	return subnet, nil
+}
+
+// getAvailableSubnets returns available Subnet under SubnetSet, and creates Subnet if necessary.
+func (r *SubnetSetReconciler) getAvailableSubnet(subnetSet *v1alpha1.SubnetSet) (*v1alpha1.Subnet, error) {
```

Comment:
For SubnetSet CR, besides create NSXSubnet, to release unused NSXSubnet, Could we also consider how to release unused NSXSubnet? for example, we can take a simple way, delete NSXSubnet if no Port on them with a conservative timeout.

Reply:
https://github.com/vmware-tanzu/nsx-operator/pull/165/files#diff-959aec2764d2baff7b43c068f97bee2f29c1c0c7e8bc8554569346cdf32499ccR54

how about recycling subnet when handing deletion of subnetport?

Reply:
do you mean recycling the subnet when subnetport deleted from subnet and no ports on it immediately?  I think this might cause subnet deletion and re-creation.

Reply:
Yea, we can take this way.

Reply:
Marking as resolved as GC will be in another patch.